File: src/env.mjs
import { spawnSync } from "child_process";
export class EnvError extends Error {
    constructor(code, message) {
        super(message);
        this.name = 'EnvError';
        this.code = code;
        this.message = message;
    }
}
export function execPwshCommand(command) {
    var _a;
    const result = spawnSync('powershell', ['-c', command]);
    if (result.status !== 0) {
        throw new EnvError((_a = result.status) !== null && _a !== void 0 ? _a : 1, result.stderr.toString().trim());
    }
    return result.stdout.toString().trim();
}
export function readEnv(name, scope) {
    return execPwshCommand(`[System.Environment]::GetEnvironmentVariable("${name}", [System.EnvironmentVariableTarget]::${scope})`);
}
export function writeEnv(name, value, scope) {
    return execPwshCommand(`[System.Environment]::SetEnvironmentVariable("${name}", "${value}", [System.EnvironmentVariableTarget]::${scope})`);
}
export function addToPath(scope, ...paths) {
    const existingPath = readEnv('PATH', scope);
    const existingPathArray = existingPath.split(';').filter((p) => p);
    const newPath = [...new Set([...existingPathArray, ...paths])].join(';');
    return writeEnv('PATH', newPath, scope);
}

File: src/helpers/env.go
package helpers

import (
	"fmt"
	"os/exec"
	"strings"
	"syscall"
)

type Scope string

const (
	ScopeUser    Scope = "User"
	ScopeMachine Scope = "Machine"
)

func execPsCommand(command string) (string, error) {
	cmd := exec.Command("powershell", "-c", command)
	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}

	output, err := cmd.Output()

	if err != nil {
		return "", fmt.Errorf("powershell command failed: %v", err)
	}

	return strings.TrimSpace(string(output)), nil
}

func psEscape(s string) string {
	return strings.ReplaceAll(s, `'`, `''`)
}

func ReadEnv(scope Scope, name string) (string, error) {
	return execPsCommand(
		fmt.Sprintf(
			"[System.Environment]::GetEnvironmentVariable('%s', [System.EnvironmentVariableTarget]::%s)",
			psEscape(name),
			scope,
		),
	)
}

func WriteEnv(scope Scope, name, value string) (string, error) {
	return execPsCommand(
		fmt.Sprintf(
			"[System.Environment]::SetEnvironmentVariable('%s', '%s', [System.EnvironmentVariableTarget]::%s)",
			psEscape(name),
			psEscape(value),
			scope,
		),
	)
}

func AddToEnvPath(scope Scope, paths ...string) (string, error) {
	existingPath, err := ReadEnv(scope, "PATH")
	if err != nil {
		return "", err
	}

	existingPathArray := strings.Split(existingPath, ";")
	var filteredPaths []string
	for _, p := range existingPathArray {
		if p != "" {
			filteredPaths = append(filteredPaths, p)
		}
	}

	pathSet := make(map[string]bool)
	var uniquePaths []string

	for _, p := range filteredPaths {
		if !pathSet[p] {
			pathSet[p] = true
			uniquePaths = append(uniquePaths, p)
		}
	}

	for _, p := range paths {
		if !pathSet[p] {
			pathSet[p] = true
			uniquePaths = append(uniquePaths, p)
		}
	}

	newPath := strings.Join(uniquePaths, ";")
	return WriteEnv(scope, "PATH", newPath)
}
